dev/src/bundle.ts

```diff
@@ -0,0 +1,184 @@
+   * Both the documents data and the query read time will be included in the bundle.
+   *
+   * @param {DocumentSnapshot | string=} documentOrName A document snapshot to add or a name of a query.
+   * @param {Query?=} querySnapshot A query snapshot to add to the bundle, if provided.
+   * @returns {BundleBuilder} This instance.
+   *
+   * @example
+   * let bundle = firestore.bundle('data-bundle');
+   * const bundleStream =
+   *    await bundle.add(await firestore.doc('abc/123').get()) // Add a document
+   *                .add('coll-query', await firestore.collection('coll').get()) // Add a named query.
+   *                .stream();
+   * // Stream `bundleStream` to clients.
+   */
+  add(
+    documentOrName: DocumentSnapshot | string,
+    querySnapshot?: QuerySnapshot
+  ): BundleBuilder {
+    if (documentOrName instanceof DocumentSnapshot) {
+      this.addBundledDocument(documentOrName);
+    } else if (querySnapshot instanceof QuerySnapshot) {
+      this.addNamedQuery(documentOrName, querySnapshot);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    if (this.namedQueries.has(name)) {
+      throw new Error(`Query name conflict: ${name} is already added.`);
+    }
+
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private lengthPrefixedBuffer(payload: string): Buffer {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+    return Buffer.concat([lengthBuffer, buffer]);
+  }
+
+  stream(): NodeJS.ReadableStream {
+    const readable = new Readable({
+      objectMode: false,
+      read: () => {
+        const metadata: firestore.IBundleMetadata = {
+          id: this.bundleId,
+          createTime: this.latestReadTime.toProto().timestampValue,
+        };
+        readable.push(
+          this.lengthPrefixedBuffer(
+            JSON.stringify({metadata} as firestore.IBundleElement)
```

Comment:
I stumbled over this again and it took me a while to figure out why your code is correct.

It might be easier to parse if `lengthPrefixedBuffer` took a `firestore.IBundleElement` and invoked JSON.stringify directly.